packages/form-control/src/FormControlMixin.ts

```diff
@@ -0,0 +1,528 @@
+import { IElementInternals } from 'element-internals-polyfill';
+import { Constructor, FormControlInterface, IControlHost, Validator } from './types';
+
+export function FormControlMixin<T extends Constructor<HTMLElement & IControlHost>>(SuperClass: T) {
+  class FormControl extends SuperClass {
+    /** Wires up control instances to be form associated */
+    static get formAssociated() {
+      return true;
+    }
+
+    /**
+     * A list of Validator objects that will be evaluated when a control's form
+     * value is modified or optionally when a given attribute changes.
+     *
+     * When a Validator's callback returns false, the entire form control will
+     * be set to an invalid state.
+     */
+    static get formControlValidators(): Validator[] {
+      return [];
+    }
```

Comment:
I'm not sure the question I had about this has been answered yet. Being a `static` would I need to make two custom inputs if one of them required a min-length and another enforced a max-length? I've done no testing on this, but from the outside `static` tends to mean that.

Reply:
No, those tend to key off of attributes and can use properties. For required, for example, the validator is 

```typescript
export const requiredValidator: Validator = {
  attribute: 'required',
  key: 'valueMissing',
  message: 'You must include a value',
  callback(instance: RequiredHost, value: any): boolean {
    let valid = true;

    if (instance.required && !value) {
      valid = false;
    }

    return valid;
  }
};
```

Based off the API here, the evaluator is run on the change of the `'required'` attribute. The callback function can look for whatever property or attribute it needs since it has access to the instance. To be more robust the check for required could be something more like
```typescript
const required = instance.required || instance.hasAttribute('required');
```

Ultimately it depends on how a given user wants to organize their validator callback per validator. If you know you always want your component to be required (like an agree to terms component), you could just as easily write:

```typescript
const alwaysRequiredValidator: Validator = {
  key: 'valueMissing',
  message: 'This field is required',
  callback(instance: HTMLElement, value: string) {
    return !!value;
  }
}
```

so it all really depends on the use case. The above won't depend or evaluate on anything except value change.